gsm extensions vignette (#1813)
Gilead-BioStats/gsm: vignettes/gsmExtensions.Rmd

  - `spec`: (required) Data table requirement for the report. Typically from `Mapped_*` or `Reporting_*` data, but occasionally from `Analytics_*` and `Reporting_*` data sources. 
  - `steps`: (required) Functions to produce the reporting output from `Mapped_*` and/or `Reporting_*` data to final output (`html`, `csv` or `shinyApp`).
   
Detailed specifications for each of these sections are provided below and links to several example module configuration files are provided at the end of this vignette (TODO: Add this). 

## Example Module Configuration YAMLs

Here are links to several sample module configuration files:

- [12 Standard gsm KRIs](https://github.com/Gilead-BioStats/gsm/tree/dev/inst/workflow/metrics) (e.g. [Adverse Event KRI Metric](https://github.com/Gilead-BioStats/gsm/tree/dev/inst/workflow/metrics/kri0001.yaml))
- [Site-](https://github.com/Gilead-BioStats/gsm/blob/dev/inst/workflow/reports/report_kri_site.yaml) and [Country-level KRI Report](https://github.com/Gilead-BioStats/gsm/blob/dev/inst/workflow/reports/report_kri_country.yaml)

> - [GSM Deep Dive Shiny App]()

Reviewer: I'm fine deleting this reference.

## `meta` Specification

The `meta` section of workflow YAML provides key metadata describing the module. It must include the following fields: 

- `Name`: Name of the reporting output.
- `Type`: "Report", "Metric" or "App"
- `ModuleID`/`MetricID`: The unique ID for the module. `ModuleID` is used for reports and apps, while `MetricID` is used for metrics.
- `Description`: A one-line description of the module specified in the workflow.
- `Details`: A more detailed description of the module specified in the workflow.

> - `Repo`: Package repo and version. Should be compatible with the `repo` parameter in `devtools::install_github()`. 

Author: yes, i am hoping that there will be a process that auto-fills/updates this field- we will want to update it with each release since all yamls should be functional at time of release, so adding it to the checklist and/or making a `version_update()` function that updates the version in Description and in all yamls (and any other locations) would make the most sense. We could even trigger that function in a GHA to prepare for release. 

I was holding off making an issue for it since it's a bit in the weeds, but I'll add it now so that we have it. It's possible that there will be a gsm.template solution here, but i'm not sure at the moment.

---
title: "gsm Extensions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{gsmExtensions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Overview

> This vignette describes how to extend `gsm` by creating new "modules", including metrics, reports and shiny apps that can be run using the standard `gsm` pipeline described in these vignettes (`vignette("DataAnalysis")`, `vignette("DataReporting")`). As shown in the `vignette("DataAnalysis")`, the `gsm` data pipeline can be used to capture a monitoring 'snapshot' for a study that includes a variety of "modules" including [metrics](https://github.com/Gilead-BioStats/gsm/tree/dev/inst/workflow/metrics/kri0001.yaml), [reports](https://github.com/Gilead-BioStats/gsm/blob/dev/inst/workflow/reports/report_kri_site.yaml) and [apps](TODO: Add this link).Some core modules are included in the `gsm` package, while others can be added as extensions. 

Reviewer: ```suggestion
This vignette describes how to extend `gsm` by creating new "modules", including metrics, reports and shiny apps that can be run using the standard `gsm` pipeline described in these vignettes (`vignette("DataAnalysis")`, `vignette("DataReporting")`). As shown in the `vignette("DataAnalysis")`, the `gsm` data pipeline can be used to capture a monitoring 'snapshot' for a study that includes a variety of "modules" including [metrics](https://github.com/Gilead-BioStats/gsm/tree/dev/inst/workflow/metrics/kri0001.yaml) and [reports](https://github.com/Gilead-BioStats/gsm/blob/dev/inst/workflow/reports/report_kri_site.yaml). Some core modules are included in the `gsm` package, while others can be added as extensions. 
```

This vignette provide detailed specifications for creating new modules, a description of the directory structure for the `yaml` workflows that comprise a module pipeline, and links to resources that can be used to configure study-level gsm pipelines that utilize these `gsm` extensions.

# Module Configuration

`gsm` modules will typically be part of an R package and have a YAML configuration file - `{module_name}.yaml` - saved under `\inst\workflow\4_modules`. Module config files include 3 key properties: 

  -  `meta`: (required) Report metadata used for centralized library and fields referenced in `steps` section of workflow.

>     - This is not required in the `{report_name}.yaml` if the report is entirely comprised of `metric`s that have their own `yaml` configuration files. 

Reviewer: I don't fully understand this line. Delete?

Author: yeah- this no longer applies. deleting.

  - `spec`: (required) Data table requirement for the report. Typically from `Mapped_*` or `Reporting_*` data. 
  - `steps`: (required) Functions to produce the reporting output from `Mapped_*` and/or `Reporting_*` data to final output (`html`, `csv` or `shinyApp`).
   

> Detailed specifications for each of these sections are provided below and links to several example module configuration files are provided at the end of this vignette (TODO: Add this). 

Reviewer: ```suggestion
Detailed specifications for each of these sections are provided below. 
```

Additional `meta` header required fields for `{gsm}` **metrics**:

- `GroupLevel`: The level at which the metric is calculated. Common values: `Site`, `Country`.
- `Abbreviation`: Abbreviation of the metric.
- `Metric`: Full name of the metric.
- `Numerator`: The numerator of the metric.
- `Denominator`: The denominator of the metric.
- `AnalysisType`: The analysis type. Common values: `rate`, `binary`
- `Model`: The model used to calculate the metric. Common values: `Normal Approximation`, `Exact`, `Bayesian`.
- `Score`: The score used to calculate the metric. Common values: `Z-Score`, `Adjusted Z-Score`, `P-Value`.
- `Threshold`: The thresholds needed to flag.

### `meta` example

A simple `meta` section for a report might look like this:
```
  File: report_example.yaml
  ID: example_report
  Name: An example report
  Description: A report that is an example
  Type: Report
  Repo: gsm.example v1.0.0
  Status: Qualified
  Permission: Users
  Outputs: An html report
  ExampleURL: https://gilead-biostats.github.io/gsm.example/example_report.html
```

## `spec` Specification

> The `spec` section of the workflow YAML specifies the data requirements for the workflow, including the data tables that are required and the columns that are needed from each table. The `gsm` data pipeline is designed to be highly customizatble, but for the purposes of this vignette, we will assume usage of the standard `gsm` data model described in the [Data Model vignette](). For this standard use case, modules will pull primarily from the "Mapped" and "Reporting" data layers. 

Reviewer: add vignette link

The `Mapped` data layer is also available for use in reports and apps. Most typically, mapped data is used to drill down from high-level metric findings (e.g. "Site 5 has an elevated AE rate relative to other studies") to site- or participant- level details (e.g. "Participant 00016 from Site 5 had 5 AEs and 3 SAEs reported in the last 3 months."). For example, the [Deep Dive app]() includes both Reporting and Mapped data in its `spec`. Here is a representative exerpt from the `spec`:

```
spec:
  Reporting_Results: 
    _all:
      required: true
  Mapped_AE:
    subjid:
      required: false
      type: character
    aeterm:
      required: false
      type: character 
    aesev: 
      required: false
      type: character
    ...
```

## `steps` Specification

Finally, each module YAML config should have a `steps` property that describes in detail how the module is run. The `steps` section is a list of functions that are run in sequence to produce the final output. Each item in `steps` has the following properties:

> - `name`: The name of the function to be run. This must be a function that is available in `{gsm}` package or in a package that is listed in the `repo` section of the `meta` header.

Reviewer 2: Might be worth mentioning that `=` is a valid function?

- `output`: The name of the output of the function. This is the name of the data table that is created by the function.
- `params`: A list of parameters that are passed to the function. The parameters are specific to the function that is being run. See below for more details on how to specify parameters for each function.

**Note**: It is important to note that the default behavior of the `RunWorkflow()` and `RunWorkflows()` functions is to return the *last* output in the steps section of the workflow. therefore, each `yaml` file- regardless of which directory it is in- should only produce one output, whether that be a data table, list, html output, deployed shiny app, or any other object needed to produce the module output.

The `steps` is the most complex part of the module configuration and will vary greatly depending on the module type and the specific requirements of the module. `gsm` provides several functions that allow for module yaml files to be run in a standard way. See `?gsm::RunWorkflow()` for more details.

### `steps[]$params` Specification

> After processing the YAML `meta` and `spec` sections, `gsm::RunWorkflow()` calls `gsm::RunStep()` for each step in the `steps` section of the YAML. The `params` section of each step is passed to `RunStep()` as a list of parameters along with a copy of the metadata header (`lMeta`) and any data (`lData`). `RunStep()` then parses the list of `params` using the logic described below and passes the parsed parameters to the function specified in the `name` field of the step.
> 
> To parse the `params` list, each item is split in to `paramName` and `paramVal`. For example, `lData: Mapped_AE` is parsed to `paramName = 'lData'` and `paramVal = 'Mapped_AE'` and the following rules are applied in order: 
> 
> 1. If `paramVal` is equal to "lMeta", the the full `lMeta` object is passed to the function (for the given `paramName`). 
> 2. If `paramVal` is equal to  "lData", the full `lData` object is passed to the function. 
> 3. If `paramVal` is found in `names(lMeta)`, that property is pulled from `lMeta` (e.g. `lMeta${paramVal}`) and passed to the function.
> 4. If `paramVal` is found in `names(lData)`, that property is pulled from `lData` (e.g. `lData${paramVal}`) and passed to the function.
> 5. Otherwise `paramVal` is passed to the function as a string. 

Reviewer: Let's just refer to the function docs for the detailed parsing rules. 

```suggestion
After processing the YAML `meta` and `spec` sections, `gsm::RunWorkflow()` calls `gsm::RunStep()` for each step in the `steps` section of the YAML. The `params` section of each step is passed to `RunStep()` as a list of parameters along with a copy of the metadata header (`lMeta`) and any data (`lData`). `RunStep()` then parses the list of `params` by passing data from `lMeta` and `lData` when appropriate - see `?RunStep` for a detailed of how parameter values are populated. Finally, the parsed parameters are passed to the function specified in the `name` field of the step. 
```

    params:
      df: Reporting_Results
      strQuery: "SELECT * FROM df WHERE GroupLevel == 'Site'"
  - name: RunQuery
    output: Reporting_Metrics_Site
    params:
      df: Reporting_Metrics
      strQuery: "SELECT * FROM df WHERE GroupLevel == 'Site'"
  - name: MakeCharts
    output: lCharts_Site
    params:
      dfResults: Reporting_Results_Site
      dfGroups: Reporting_Groups
      dfBounds: Reporting_Bounds
      dfMetrics: Reporting_Metrics_Site
  - name: Report_KRI
    output: lReport 
    params:
      lCharts: lCharts_Site
      dfResults: Reporting_Results_Site
      dfGroups: Reporting_Groups
      dfMetrics: Reporting_Metrics_Site
```

# Directory Structure for Workflows 

Each extension that produces report(s) will have a `workflow` directory in the `inst` of the package that follows a standard structure. This directory will contain 4 folders in which to store the `yaml` workflow files that map data, perform analysis, produce reporting data, and generate the output of a module. Each module output requires it's own unique `yaml` in the `4_modules` folder, which will take inputs generated from the previous three directories.

### `/1_mappings`

> The mappings folder contains all of the mappings from `Raw_*` data to `Mapped_*` data. Each file within this directory is to be named for the data table it is creating, minus the `Mapped_` suffix. The `yaml`s will contain the three required sections, which are discussed in detail in the `Module Configuration` section above. The `yaml`s in this folder will be combined via `CombineSpecs()` to create a master spec that defines all necessary tables and columns for the module(s) in this package.

Reviewer 2: ```suggestion
The mappings folder contains all of the mappings from `Raw_*` data to `Mapped_*` data. Each file within this directory is to be named for the data table it is creating, minus the `Mapped_` suffix. The `yaml`'s will contain the three required sections, which are discussed in detail in the `Module Configuration` section above. The `yaml`'s in this folder will be combined via `CombineSpecs()` to create a master spec that defines all necessary tables and columns for the module(s) in this package.
```

`yaml`s?
`yamls`?
`yaml`'s
`yaml`-s

Idk which looks best?

Author: yeah, this is something i've been thinking about as i was writing this up. i think either `yamls` or just not adding ticks at all and just typing it all in plain text make the most sense to me.

    n_changes:
      required: true
      type: numeric
  Mapped_SUBJ:
    subjid:
      required: true
      type: character
    subject_nsv:
      required: true
      type: character
steps:
  # Merge [ subjid ] onto EDC domains.
  - output: Temp_SubjectLookup
    name: select
    params:
      .data: Mapped_SUBJ
      subjid: subjid
      subject_nsv: subject_nsv
  - output: Mapped_DATACHG
    name: left_join
    params:
      x: Raw_DATACHG
      "y": Temp_SubjectLookup
      by: subject_nsv
```

### `/2_metrics`

The metrics directory contains all of the workflows that perform analysis steps, converting mapped data into metrics that are displayed in a report. In the case of `{gsm}`, these metrics are the 12 Key Risk Indicators, calculated at both the site- and country-level, that are discussed in the Data Analysis Step-by-Step Vignette. Each `yaml` in this file produces a list of analysis data tables that capture the formatted input table, the transformed table, the flagged table, and the summary table. In general, these `yaml`s should at least provide a summary table that contains statistics about the metric at the specified level of aggregation.

> Examples of these yamls can be found above in the `Module Configuration` section, as well as in the Data Analysis Vignette.

Reviewer 2: ```suggestion
Examples of these `yaml`s can be found above in the `Module Configuration` section, as well as in the (`vignette("DataAnalysis")`) vignette.
```

i don't mind the lack of ticks plural form of yamls actually, maybe thats a good option, but highlighting for consistency and link to the appropriate vignette

### `/3_reporting`

The reporting directory is intended to hold all of the workflows that produce the data that is required for the module outputs. This typically requires the stacking of analysis data from all of the relevant metrics into a single `results` data frame that can be surfaced in a report, or multiple reports. Additionally, any further information that must be taken from the analysis output, such as study/site/group/metric metadata and supporting statistics will be constructed through workflows in this folder.

> Examples of these `yaml`s can be found above in the `Module Configuration` section, as well as in the Data Reporting Vignette.

Reviewer 2: ```suggestion
Examples of these `yaml`s can be found above in the `Module Configuration` section, as well as in the Data Reporting Vignette.
```

add link to datareporting vignette?
